### PR TITLE
Use 3D optimization problem.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 1]
```diff
@@ -186,13 +188,13 @@ void SparsePoseGraph::ComputeConstraint(const mapping::NodeId& node_id,
...
     // search window.
-    const transform::Rigid2d initial_relative_pose =
+    const transform::Rigid3d initial_relative_pose =
         optimization_problem_.submap_data().at(submap_id).pose.inverse() *
         optimization_problem_.node_data().at(node_id).pose;
     constraint_builder_.MaybeAddConstraint(
         submap_id, submap_data_.at(submap_id).submap.get(), node_id,
         trajectory_nodes_.at(node_id).constant_data.get(),
-        initial_relative_pose);
+        transform::Project2D(initial_relative_pose));
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: `transform::Project2D(initial_relative_pose * transform::Rigid3d::Rotation(trajectory_nodes_.at(node_id).constant_data->gravity_alignment.inverse())`

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 2]
```diff
@@ -221,29 +223,26 @@ void SparsePoseGraph::ComputeConstraintsForScan(
   CHECK_EQ(submap_ids.size(), insertion_submaps.size());
   const mapping::SubmapId matching_id = submap_ids.front();
   const auto& constant_data = trajectory_nodes_.at(node_id).constant_data;
-  const transform::Rigid2d pose = transform::Project2D(
+  const transform::Rigid3d pose =
       constant_data->initial_pose *
-      transform::Rigid3d::Rotation(constant_data->gravity_alignment.inverse()));
-  const transform::Rigid2d optimized_pose =
+      transform::Rigid3d::Rotation(constant_data->gravity_alignment.inverse());
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Remove this line?

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 3]
```diff
@@ -429,10 +428,8 @@ void SparsePoseGraph::AddNodeFromProto(const int trajectory_id,
         constant_data->gravity_alignment.inverse());
     optimization_problem_.AddTrajectoryNode(
         node_id.trajectory_id, constant_data->time,
-        transform::Project2D(constant_data->initial_pose *
-                             gravity_alignment_inverse),
-        transform::Project2D(pose * gravity_alignment_inverse),
-        constant_data->gravity_alignment);
+        constant_data->initial_pose * gravity_alignment_inverse,
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Remove the `gravity_alignment_inverse` here and on the next line?

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 4]
```diff
@@ -477,7 +474,7 @@ void SparsePoseGraph::RunOptimization() {
       const mapping::NodeId node_id = node_it->id;
       auto& node = trajectory_nodes_.at(node_id);
       node.pose =
-          transform::Embed3D(node_it->data.pose) *
+          node_it->data.pose *
           transform::Rigid3d::Rotation(node.constant_data->gravity_alignment);
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Remove this line?

File: cartographer/mapping_2d/sparse_pose_graph.cc (hunk 3 above)
Comment (cschuet) [reply cartographer/mapping_2d/sparse_pose_graph.cc]: Done

**Comment by SirVer** [comment]
@cschuet I suggest closing this PR since we have no immediate way forward here.

**Pull request closed without merging**